engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -184,20 +215,79 @@ class WorkflowActor(workflowId: WorkflowId,
+    case Event(MetadataServiceFailureResponse(error), _) =>
+      // Do something useful here??
+      log.warning(s"$tag Received MetadataServiceFailure")
+      stay
+    case Event(MetadataActionFailureResponse(error, action), _) =>
+      // Do something useful here??
+      log.warning(s"$tag Received MetadataServiceFailure")
+      stay
     case Event(AbortWorkflowCommand, WorkflowActorData(Some(actor), _)) =>
       actor ! EngineLifecycleActorAbortCommand
       goto(AbortingWorkflowState)
     case unhandledMessage =>
       log.warning(s"$tag received an unhandled message $unhandledMessage in state $stateName")
       stay
   }
+
+  onTransition {
+    case fromState -> toState =>
+      log.info(s"$tag transitioning from $fromState to $toState")
+      // This updates the workflow status
+      pushCurrentStateToMetadataService(toState.workflowState) onFailure {
+        case reason: Throwable => self ! MetadataServiceFailureResponse(reason)
+      }
+  }
+
+  onTransition {
+    case oldState -> terminalState if terminalState.terminal =>
+      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
+      // Add the end time of the workflow in the MetadataService
+      val metadataEventMsg = MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.EndTime), MetadataValue(DateTime.now.toString))
+      serviceRegistryActor.publishToMetadataService(metadataEventMsg) map handleMetadataServiceResponse recover {
+        case reason: Throwable =>
+          self ! MetadataServiceFailureResponse(reason)
+      } andThen {
+        case _ =>
+          terminalState match {
+            case WorkflowSucceededState =>
+              context.parent ! WorkflowSucceededResponse(workflowId)
+            case WorkflowFailedState =>
+              context.parent ! WorkflowFailedResponse(workflowId, oldState, Seq.empty[Throwable])
+            case WorkflowAbortedState =>
+              context.parent ! WorkflowAbortedResponse(workflowId)
+            case unknownState => log.warning(s"tag $unknownState is an unhandled terminal state!")
+          }
+      }
+  }
+
+  private def pushWfNameAndInputsToMetadataService(workflowDescriptor: EngineWorkflowDescriptor): Future[Unit] = {
+    val inputMetadataEvents = workflowDescriptor.backendDescriptor.inputs.map { case (k, v) =>
+      MetadataEvent(MetadataKey(workflowId, None, s"${WorkflowMetadataKeys.Inputs}.$k"), MetadataValue(v.toWdlString))
+    }
+    val metadataEventMsgs = List(MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.Name), MetadataValue(workflowDescriptor.name))) ++ inputMetadataEvents
+    Future.sequence(metadataEventMsgs map serviceRegistryActor.publishToMetadataService).map {
+      _ foreach handleMetadataServiceResponse
+    }
+  }
+
+  // Update the current State of the Workflow (corresponding to the FSM state) in the Metadata service
+  private def pushCurrentStateToMetadataService(workflowState: WorkflowState): Future[Unit] = {
```

Comment:
I think instead of defining private methods inside dedicated workflow actor, meta data related behaviors can be moved to metadata actor, that would be more appropriate than growing workflow actor, However follow SRP (Single Responsibility principle), it looks like it is breaking it right now.